### PR TITLE
Add more posix-native misc things.

### PR DESCRIPTION
**Pull request by BillyONeal**
This PR is part of the std::filesystem removal-on-POSIX project.

* rename
* is_directory
* is_regular_file
* is_empty
* create_directory
* create_directories
* create_symlink
* create_directory_symlink
* status
* symlink_status

### CODE REVIEW

File: src/vcpkg/base/files.cpp [hunk 1]
```diff
@@ -2208,35 +2259,197 @@ namespace vcpkg
...
+                }
+
+                last_mkdir_created = attempt != 0;
+                if (next_slash == last)
+                {
+                    return last_mkdir_created;
+                }
+
+                this_create.push_back('/');
+                first = std::find_if_not(next_slash, last, is_slash);
```
Comment (strega-nil-ms) [inline src/vcpkg/base/files.cpp:2403]: Could you add a comment here about this existing for purposes of multiple slashes in a row?

File: src/vcpkg/base/files.cpp [hunk 2]
```diff
@@ -2208,35 +2259,197 @@ namespace vcpkg
...
+            }
+
+#endif // _WIN32
         }
+
+#if !defined(_WIN32)
+        static void posix_create_symlink(const Path& to, const Path& from, std::error_code& ec)
+        {
+            if (symlink(to.c_str(), from.c_str()) == 0)
```
Comment (strega-nil-ms) [inline src/vcpkg/base/files.cpp:2412]:
```suggestion
            if (::symlink(to.c_str(), from.c_str()) == 0)
```

File: src/vcpkg/base/files.cpp [hunk 3]
```diff
@@ -2208,35 +2259,197 @@ namespace vcpkg
...
+#else  // ^^^ _WIN32 // !_WIN32 vvv
+            posix_create_symlink(to, from, ec);
+#endif // _WIN32
         }
         virtual void create_hard_link(const Path& to, const Path& from, std::error_code& ec) override
         {
+#if defined(_WIN32)
             stdfs::create_hard_link(to_stdfs_path(to), to_stdfs_path(from), ec);
+#else  // ^^^ _WIN32 // !_WIN32 vvv
+            if (link(from.c_str(), to.c_str()) == 0)
```
Comment (strega-nil-ms) [inline src/vcpkg/base/files.cpp:2444]:
```suggestion
            if (::link(from.c_str(), to.c_str()) == 0)
```

File: src/vcpkg/base/files.cpp [hunk 4]
```diff
@@ -2260,15 +2473,51 @@ namespace vcpkg
 
         virtual FileType status(const Path& target, std::error_code& ec) const override
         {
+#if defined(_WIN32)
             auto result = stdfs::status(to_stdfs_path(target), ec);
             translate_not_found_to_success(ec);
             return convert_file_type(result.type());
+#else  // ^^^ _WIN32 // !_WIN32 vvv
+            struct stat s;
+            if (stat(target.c_str(), &s) == 0)
```
Comment (strega-nil-ms) [inline src/vcpkg/base/files.cpp:2482]:
```suggestion
            if (::stat(target.c_str(), &s) == 0)
```

File: src/vcpkg/base/files.cpp [hunk 5]
```diff
@@ -2260,15 +2473,51 @@ namespace vcpkg
...
         }
         virtual FileType symlink_status(const Path& target, std::error_code& ec) const override
         {
+#if defined(_WIN32)
             auto result = stdfs::symlink_status(to_stdfs_path(target), ec);
             translate_not_found_to_success(ec);
             return convert_file_type(result.type());
+#else  // ^^^ _WIN32 // !_WIN32 vvv
+            struct stat s;
+            if (lstat(target.c_str(), &s) == 0)
```
Comment (strega-nil-ms) [inline src/vcpkg/base/files.cpp:2506]:
```suggestion
            if (::lstat(target.c_str(), &s) == 0)
```

**Review by strega-nil-ms — APPROVED** [review]
Minor code style changes, approved :)

File: src/vcpkg/base/files.cpp [hunk 6]
```diff
@@ -2208,35 +2259,197 @@ namespace vcpkg
...
+            return S_ISDIR(s.st_mode);
+#endif // ^^^ !_WIN32
         }
         virtual bool is_regular_file(const Path& target) const override
         {
+#if defined(_WIN32)
             return stdfs::is_regular_file(to_stdfs_path(target));
+#else  // ^^^ _WIN32 // !_WIN32 vvv
+            struct stat s;
+            if (stat(target.c_str(), &s) != 0)
```
Comment (ras0219-msft) [inline src/vcpkg/base/files.cpp:2280]:
```suggestion
            if (::stat(target.c_str(), &s) != 0)
```

File: src/vcpkg/base/files.cpp [hunk 7]
```diff
@@ -2208,35 +2259,197 @@ namespace vcpkg
 
         virtual bool is_directory(const Path& target) const override
         {
+#if defined(_WIN32)
             return stdfs::is_directory(to_stdfs_path(target));
+#else  // ^^^ _WIN32 // !_WIN32 vvv
+            struct stat s;
+            if (stat(target.c_str(), &s) != 0)
```
Comment (ras0219-msft) [inline src/vcpkg/base/files.cpp:2266]:
```suggestion
            if (::stat(target.c_str(), &s) != 0)
```

File: src/vcpkg/base/files.cpp [hunk 8]
```diff
@@ -2011,7 +2051,7 @@ namespace vcpkg
             std::vector<Path> result;
             get_files_non_recursive_impl(result, dir, ec, [](PosixDType dtype, const Path& p) {
                 return dtype == PosixDType::Directory ||
-                       (dtype == PosixDType::Unknown && posix_is_directory(p.c_str()));
+                       (dtype == PosixDType::Unknown && lstat_is_directory(p.c_str()));
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/base/files.cpp]: Good call, should be stat here: http://eel.is/c++draft/fs.class.directory.entry#fs.dir.entry.obs-9

That also means enumeration might need to do both :(

**Pull request merged**